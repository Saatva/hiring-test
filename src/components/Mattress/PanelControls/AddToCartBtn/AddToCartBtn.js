

import React, { useState } from 'react';

import addToCartBtnClasses from './AddToCartBtn.module.scss';

const AddToCartBtn = (props) => {
    const [disabled, setDisabled] = useState(false);

    const clickHandler = () => {
        props.itemAdded();
        setDisabled(true);
        setTimeout(() => setDisabled(false));
    }

    return (
        <button
            className={addToCartBtnClasses.AddToCartBtn}
            disabled={disabled}
            onClick={clickHandler}>Add to Cart</button>
    );
};

export default AddToCartBtn;
