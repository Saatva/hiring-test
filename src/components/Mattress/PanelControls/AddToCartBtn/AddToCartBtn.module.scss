.AddToCartBtn {
    background-color: #d5aa63;
    border: 1px solid hsl(35, 100%, 50%);
    border-radius: 0;
    border: 1px solid #4d4d4d;
    color: #fff;
    cursor: pointer;
    font-size: 17px;
    font-family: 'Questrial', sans-serif;
    font-weight: 400;
    line-height: 1;
    margin: 2rem 0;
    outline: none;
    padding: 0.8rem 1.5rem;
    border: 1px solid #4d4d4d;
    border-color: #d5aa63;
    transition: background-color .25s ease;
}

.AddToCartBtn:hover,
.AddToCartBtn:focus {
  background-color: hsl(35, 100%, 45%);
}

.AddToCartBtn:not(:disabled) {
  animation: animateAddToCartButton 0.3s linear;
}

@keyframes animateAddToCartButton {
  0% {
      transform: scale(1);
  }
  60% {
      transform: scale(1.1);
  }
  100% {
      transform: scale(1);
  }
}