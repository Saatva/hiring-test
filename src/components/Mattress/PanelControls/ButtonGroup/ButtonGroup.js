import React, { useState, useRef, useEffect } from 'react';

import buttonGroupClasses from './ButtonGroup.module.scss';

const ButtonGroup = (props) => {
    const [clickedBtnIndex, setClickedBtnIndex] = useState(0);
    const btnGroupElm = useRef(null);
    const btnIndxRef = useRef(0);
    const buttonsRef = useRef([]);

    const clickHandler = (mattress, i) => {
        props.buttonClicked(mattress);
        setClickedBtnIndex(i)
        btnIndxRef.current = i;
    };

    const buildBtnGroup = (mattresses) => {
        return (
            mattresses.map((mattress, i) => (
                <button
                    key={i}
                    ref={el => buttonsRef.current[i] = el}
                    name={mattress.name}
                    onClick={() => clickHandler(mattress, i)}
                >
                    {mattress.name}
                </button>
            ))
        )
    };

    const handleClickCallback = e => {
        if (btnGroupElm.current.contains(e.target)) {
            return;
        }

        if (clickedBtnIndex === btnIndxRef.current)
            setClickedBtnIndex(-1)
        else
            setClickedBtnIndex(clickedBtnIndex)
    };

    useEffect(() => {
        buttonsRef.current[btnIndxRef.current].focus();
        setClickedBtnIndex(btnIndxRef.current)
    }, [clickedBtnIndex]);

    useEffect(() => {
        buttonsRef.current[0].focus();
        document.addEventListener("mousedown", handleClickCallback);

        return () => {
            document.removeEventListener("mousedown", handleClickCallback);
        };
    }, []);

    return (
        <div className={buttonGroupClasses.ButtonGroup} ref={btnGroupElm}>
            {buildBtnGroup(props.mattresses)}
        </div>
    )
};

export default ButtonGroup;
