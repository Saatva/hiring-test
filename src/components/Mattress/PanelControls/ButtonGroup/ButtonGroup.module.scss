.ButtonGroup {
  display: flex;
  width: 100%;
  button {
    padding: 10px 24px;
    cursor: pointer;
    float: left;
    font-family: 'Questrial', sans-serif;
    font-size: 1rem;
    font-weight: 400;
    line-height: 1;
    outline: none;
    color: #6b6257;
    background-color: #fff;
    border: 1px solid #d2d2d2;
    text-align: center;
    white-space: nowrap;
    flex: 1;
    &:not(:last-child) {
      border-right: none;
    }
    &:hover {
      background-color: hsl(0, 0%, 95%);
    }
    &:focus,
    &:active {
      color: #fff;
      background-color: #a6a19a;
      border-color: #d2d2d2;
    }
  }
}
