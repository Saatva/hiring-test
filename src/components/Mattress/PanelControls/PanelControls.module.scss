.PanelControls {
    background-color: hsl(0, 0%, 95%);
    display: flex;
    flex-direction: column;
    align-items: left;
    justify-content: center;
    margin: auto;
    padding: 20px;
    width: 100%;
}

.PanelControls h1 {
  font-family: 'Source Serif Pro', serif;
}

.PanelControls p, label {
  font-family: 'Questrial', sans-serif;
}

.PanelControls label {
  padding-bottom: 4px;
}

.PanelControls div {
  display: flex;
  justify-content: space-between;
}
