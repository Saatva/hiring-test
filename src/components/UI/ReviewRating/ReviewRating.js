import React, { useState, useEffect } from 'react';

import reviewRatingClasses from './ReviewRating.module.scss';

const ReviewRating = (props) => {
    const [rating, setRating] = useState(0);

    useEffect(() => {
        setRating(props.rating);
    }, [props.rating]);

    return (
        <div className={reviewRatingClasses.ReviewRating}>
            {[0, 1, 2, 3, 4, 5].map((star, i) => {
                return (
                    <span key={i}>
                        {rating < star ?
                            0 < star ? "☆" : "★"
                            :
                            "★"
                        }
                    </span>
                );
            })}
        </div>
    );
};


export default ReviewRating;