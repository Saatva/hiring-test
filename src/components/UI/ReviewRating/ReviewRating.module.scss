.ReviewRating {
    font-size: 2em; 
    color: #d5aa63;
}