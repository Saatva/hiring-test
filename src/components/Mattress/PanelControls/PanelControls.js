import React, { useState } from 'react';

import panelControlsClasses from './PanelControls.module.scss';
import ButtonGroup from './ButtonGroup/ButtonGroup';
import AddToCartBtn from './AddToCartBtn/AddToCartBtn';
import ReviewRating from '../../UI/ReviewRating/ReviewRating';

const PanelControls = (props) => {
    const [selectedMattress, setSelectedMattress] = useState(props.mattresses[0]);

    const selectedMattressHandler = (mattress) => {
        setSelectedMattress(mattress)
        props.selectedMattress(mattress);
    };

    return (
        <div className={panelControlsClasses.PanelControls}>
            <h1>Choose Your Mattress</h1>

            <label>SELECT MATTRESS TYPE</label>

            <ButtonGroup buttonClicked={selectedMattressHandler} mattresses={props.mattresses} />

            <div>
                <p>{selectedMattress.name} Mattress</p>
                <p>${selectedMattress.price}</p>
            </div>

            <div>
                <ReviewRating rating={selectedMattress.reviewRating}/>
            </div>

            <AddToCartBtn itemAdded={props.itemAdded} />
        </div>
    );
};

export default PanelControls;
