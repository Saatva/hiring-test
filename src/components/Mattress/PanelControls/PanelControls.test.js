import React from 'react';
import PanelControls from './PanelControls';
import ButtonGroup from './ButtonGroup/ButtonGroup';
import ReviewRating from './../../UI/ReviewRating/ReviewRating';
import AddToCartBtn from './AddToCartBtn/AddToCartBtn';
import { configure, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

configure({ adapter: new Adapter() });


describe('<PanelControls />', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<PanelControls mattresses={[{name: 'Classic', price: 120}]} />);
    });

    it('should match the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render <ButtonGroup />', () => {
        expect(wrapper.find(ButtonGroup)).toHaveLength(1);
    })

    it('should render <ReviewRating />', () => {
        expect(wrapper.find(ReviewRating)).toHaveLength(1);
    })

    it('should render <AddToCartBtn />', () => {
        expect(wrapper.find(AddToCartBtn)).toHaveLength(1);
    })

    it('should contain the correct HTML elements with data />', () => {
        expect(wrapper.contains(<p>Classic Mattress</p>)).toEqual(true);
        expect(wrapper.contains(<p>$120</p>)).toEqual(true);
    })
})