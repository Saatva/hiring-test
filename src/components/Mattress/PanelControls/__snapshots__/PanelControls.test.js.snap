// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PanelControls /> should match the snapshot 1`] = `"<div class=\\"PanelControls\\"><h1>Choose Your Mattress</h1><label>SELECT MATTRESS TYPE</label><div class=\\"ButtonGroup\\"><button name=\\"Classic\\">Classic</button></div><div><p>Classic Mattress</p><p>$120</p></div><div><div class=\\"ReviewRating\\"><span>★</span><span>☆</span><span>☆</span><span>☆</span><span>☆</span><span>☆</span></div></div><button class=\\"AddToCartBtn\\">Add to Cart</button></div>"`;

exports[`Link changes the class when hovered 1`] = `
<div
  className="PanelControls"
>
  <h1>
    Choose Your Mattress
  </h1>
  <label>
    SELECT MATTRESS TYPE
  </label>
  <div
    className="ButtonGroup"
  >
    <button
      name="test"
      onClick={[Function]}
    >
      test
    </button>
  </div>
  <div>
    <p>
      test
       Mattress
    </p>
    <p>
      $
    </p>
  </div>
  <div>
    <div
      className="ReviewRating"
    >
      <span>
        ★
      </span>
      <span>
        ☆
      </span>
      <span>
        ☆
      </span>
      <span>
        ☆
      </span>
      <span>
        ☆
      </span>
      <span>
        ☆
      </span>
    </div>
  </div>
  <button
    className="AddToCartBtn"
    disabled={false}
    onClick={[Function]}
  >
    Add to Cart
  </button>
</div>
`;
